docs(ui-coverage): rework "Work with AI agents" for accuracy, value, and clarity (#6747)
cypress-io/cypress-documentation: docs/ui-coverage/work-with-ai-agents.mdx

# Work with AI agents

> UI Coverage gives your AI agent the grounded context it needs to improve your tests. For every recorded Cypress run, it reports which interactive elements and links your tests exercised, assigns a coverage score to each view, and names the exact untested controls to go after, all built from what really ran in the browser during [Test Replay](/cloud/features/test-replay). Working from that report, an agent can find real coverage gaps and write tests that target them, rather than guessing at your app's structure.

Reviewer: Suggesting to break up the long sentence here, and frame up front that missing coverage is pre-filtered by a configuration that represents your intent. A common early question for UIC is "we already know we don't want to test everything with Cypress, how do we scope it down to what is needed for our teams and projects?". 

```suggestion
UI Coverage gives your AI agent the grounded context it needs to improve your tests. For every recorded Cypress run, it reports which interactive elements and links your tests exercised, assigns a coverage score to each view, and names the exact untested controls to go after. It's all built from what really ran in the browser during [Test Replay](/cloud/features/test-replay), and filtered through your your configuration to focus on what already matters to you. Working from that report, an agent can find real coverage gaps and write tests that target them, rather than guessing at your app's structure.
```

## What an AI agent can do with UI Coverage

Connected to your UI Coverage report, an agent can turn coverage gaps into concrete testing work:

> - **Turn "coverage dropped" into a work list**: which views regressed, which elements are newly untested, and where the risk is concentrated.

Reviewer: ```suggestion
- **Turn "coverage dropped" into a work list**: which views regressed, which elements are newly untested, and where the risk is concentrated.
- **Prioritize missing scenarios**: Comparing missing coverage with your currently-tested user-journeys can put the gaps in context.
```

- **Draft tests for untested elements**, and suggest the existing spec file that's the best home for them.
- **Correlate gaps with ownership**: group untested areas by team, feature, or route so the right people get the right list.

> - **Backfill incrementally**: sequence the work from highest-risk to lowest instead of dumping every control on every page.

Reviewer: ```suggestion
- **Backfill incrementally**: sequence the work from highest-risk to lowest instead of dumping every control on every page.
- **Configure with intent**: Generate the configuration needed to ignore or group elements as needed to remove a gap.
```

Your team still decides what "enough coverage" means for each area. The agent handles the triage, drafting, and bookkeeping around that judgment.

## Quickstart

Three steps to your first useful result:

1. **Connect the MCP.** Add Cypress Cloud MCP to your AI client by following [Configuring the Cloud MCP](/cloud/integrations/cloud-mcp#Configuring-the-Cloud-MCP).
2. **Ask for the riskiest view.** Prompt: _"Using Cypress Cloud, pull the UI Coverage report for the latest run on this branch and list the untested elements on the riskiest view."_
3. **Act on one gap.** Have the agent [draft a test](#Prompts-to-plan-and-draft-tests) for one of those elements, review the diff, and run it.

The rest of this page deepens each step: better prompts, report tuning, and how to keep the score honest.

## Query UI Coverage with Cypress Cloud MCP

> [Cypress Cloud MCP](/cloud/integrations/cloud-mcp), Cypress Cloud's Model Context Protocol server, lets compatible agents read UI Coverage results for a recorded run through structured tools, so they work from your real report. Set it up once by following [Configuring the Cloud MCP](/cloud/integrations/cloud-mcp#Configuring-the-Cloud-MCP), then prompt your agent in natural language, and it selects the right tool for you.

Reviewer: ```suggestion
Prompt your agent in natural language, and it selects the right tool for you.
```

The docs get right into the MCP above this, so we could nest more of that earlier content under the MCP section if we want. Since there are MCP, Test Gen, CLI, Results API & maybe webhooks & other AI integration points, it makes sense for MCP to be one of multiple top-level ideas discussed.

Not necessary in this PR though -- this suggestion just removes what seems like a re-introduction of the MCP.

- **Name the run** with a branch, commit, or run URL, so the agent reads the report you mean.
- **Start with the riskiest view** and work one view at a time, rather than the whole app at once.
- **Ask for a plan or a diff** before any file changes, so you review intent before code lands.
- **Use your own view names and paths.** The examples below are patterns to adapt, not literal values; swap in the views, flows, and folders from your app.

## Prompts to summarize and triage coverage

Point the agent at a run and ask it to summarize where the risk is.

<CopyPrompt
  title="Summarize coverage for a run"
  subtext="Pull the UI Coverage report for the latest run and surface the riskiest view."
  prompt={`Using Cypress Cloud, pull the UI Coverage report for the latest run on this branch. Give me the overall coverage score, the views with the most untested elements, and the specific untested elements on the riskiest view.`}
>

### Summarize coverage for a run

</CopyPrompt>

<CopyPrompt
  title="Find where new tests belong"
  subtext="List the untested elements on a view and suggest which spec should cover them."

>   prompt={`Get UI Coverage for this run URL. Pick the riskiest view, list its untested interactive elements, and suggest which existing spec file is the best place to add a test that exercises them.`}

Reviewer: ```suggestion
  prompt={`Get UI Coverage report for the latest run on this branch. Pick the riskiest view, list its untested interactive elements, and suggest which existing spec file is the best place to add a test that exercises them.`}
```

Making this more likely to succeed when pasted by removing the vague "this run URL".

>

### Find where new tests belong

</CopyPrompt>

<CopyPrompt
  title="Assess release readiness"
  subtext="Get a go/no-go on your critical flows, with the exact gaps blocking each one."

>   prompt={`Using Cypress Cloud MCP, get the UI Coverage report for the latest run on our release branch. For the views behind our most critical user flows, give me each view's coverage score and its untested interactive elements. Then give me a go/no-go: flag any critical view below our threshold (say 80%) and list the specific gaps blocking it.`}

Reviewer: ```suggestion
  prompt={`Using Cypress Cloud MCP, get the UI Coverage report for the latest run on this branch. For the views behind our most critical user flows, give me each view's coverage score and its untested interactive elements. Then give me a go/no-go: flag any critical view below our threshold (say 80%) and list the specific gaps blocking it.`}
```

  prompt={`Using Cypress Cloud MCP, get the UI Coverage report for the latest run on this branch and for the latest run on main. Compare the overall coverage scores, then diff the untested interactive elements and links between the two runs: list what became untested on this branch (regressions) and what is now covered (gains), grouped by view. For each regression, point me to the spec or app change most likely responsible.`}
>

### Diff coverage between two branches

</CopyPrompt>

<CopyPrompt
  title="Draft tests for untested elements"
  subtext="Draft focused tests for untested elements on a view, diff-first."
  prompt={`You are a Cypress test author. Using Cypress Cloud, take the untested elements on the view I'm working on from the latest run. For each, draft a focused test using our existing data-cy selectors and add it to the closest matching spec in our e2e folder. Show me a diff before writing files.`}
>

### Draft tests for untested elements

</CopyPrompt>

<CopyPrompt
  title="Author tests in natural language with cy.prompt"
  subtext="Cover untested areas by describing each flow in plain language and letting cy.prompt write the Cypress commands."

>   prompt={`Using Cypress Cloud, list the untested interactive elements on the riskiest view for the latest run. For each, add a test that uses cy.prompt to drive the flow in natural language: describe the user action to take and what to verify in plain English, and let cy.prompt generate the Cypress commands. Put the tests in our e2e folder and show me a diff before writing files.`}

Reviewer: ```suggestion
  prompt={`Using Cypress Cloud, list the untested interactive elements on the riskiest view for the latest run. For each, add a test that uses cy.prompt to drive the flow in natural language: describe the user action to take and what to verify in plain English, and let cy.prompt generate the Cypress commands. Use the Cypress Author skill, if available, to create well-formed prompt steps following best practices. Put the tests in our e2e folder and show me a diff before writing files.`}
```

Hinting to use the cypress author skill made a big difference in the quality of the prompt commands that were generated. Without the hint, the skill was skill referenced in generating the test but my Cursor session didn't use it for making well-formed prompt steps.

  prompt={`First read https://docs.cypress.io/llm/markdown/ui-coverage/configuration/overview.md for the App Quality configuration reference. Then use Cypress Cloud MCP to list the untested interactive elements for the latest run on this branch, flag third-party widgets and controls split apart by dynamic attributes, and propose App Quality Config rules (elementFilters, attributeFilters, elementGroups) to remove that noise. Show the JSON I can paste into the App Quality tab.`}
>

### Draft configuration to cut noise

</CopyPrompt>

## Keep coverage honest

Coverage measures whether your tests _touch_ an element, not whether they confirm the desired behavior. An agent told to raise the score can write tests that click everything and assert nothing, lifting the number without adding real protection.

:::caution

Require meaningful assertions, not just interactions. Ask the agent to assert the outcome of each action, and use [`allowedInteractionCommands`](/ui-coverage/configuration/allowedinteractioncommands) to require a real check (for example, listing `assert`) on elements where a bare click shouldn't count as tested.

> :::

Reviewer: ```suggestion
Coverage measures whether your tests _touch_ an element, not whether they confirm the desired behavior. An agent told to raise the score can write tests that click everything and assert nothing, so it's important to ground your coverage in the scenarios and outcomes that matter.

Ensure tests include assertions about the state of the application, and the readiness of the DOM for the next step of the text. 

:::info

For areas where assertions on non-interactive elements are critical to track, your code can use the [`data-cy-ui-interactive`](/ui-coverage/core-concepts/interactivity#Overriding-interactive-elements-with-data-cy-ui-interactive) attribute to bring those elements into your coverage report. 

Combine this with [`allowedInteractionCommands`](/ui-coverage/configuration/allowedinteractioncommands) configuration to allow assertions to count as coverage for those elements. This adds an extra layer of guidance for your agent about what must be included in your tests.

:::
```

I couldn't follow this section at first, I've adjusted this a bit to explain how to get "assertion coverage", and avoid describing interactions as not "real checks" -- usually it's the opposite, for interactive see assertions as pretty shallow and actual interactions as the real check that proves an element works.

Keep a person in the loop wherever judgment matters:

- **Intent.** Confirm the drafted test verifies the right behavior, not just that an element was interacted with.
- **What's worth testing.** Agents target trivial controls as eagerly as critical ones. You decide which gaps matter.

> - **Selectors.** When your app lacks stable `data-*` attributes, agent-written selectors can be brittle. Add attributes (or [`significantAttributes`](/ui-coverage/configuration/significantattributes)) rather than accept fragile ones, or have the agent author the test with [`cy.prompt`](/api/commands/prompt), whose selectors self-heal as your app changes.

Reviewer: ```suggestion
- **Selectors.** When your app lacks stable `data-*` attributes, agent-written selectors can be brittle. Add `data-cy` attributes rather than accept fragile ones, or have the agent author the test with [`cy.prompt`](/api/commands/prompt), whose selectors self-heal as your app changes.
```

Significant attributes config doesn't make sense here as it would not impact test stability.

> - **Intentional drops.** A coverage decrease after removing a feature is correct. Don't let an agent "fix" it.

Reviewer: If you have 100% coverage and remove a feature, you can have fewer tested elements in total on the next run, but still 100% of the available elements can be tested.

Updating this to refer to removed _tests_ instead.

```suggestion
- **Intentional drops.** A coverage decrease after removing a test is correct. Don't let an agent "fix" it by re-adding the test, but instead, filter out the elements you do not intend to cover with tests.
```

## Generate tests directly from the report

Prefer not to wire up an agent at all? Cypress Cloud can write the test for you. Select an untested element in a UI Coverage report and click **Generate test code**, and [UI Coverage Test Generation](/ui-coverage/guides/address-coverage-gaps#Generate-tests-from-UI-Coverage-reports) drafts a Cypress test that navigates to the element and interacts with it, following your existing patterns.

<DocsImage
  src="/img/ui-coverage/guides/cypress-ui-coverage-generate-test.png"
  alt="Cypress Cloud showing a generated Cypress test after clicking Generate test code for an untested link in a UI Coverage report."
/>

## Standardize across your team

These workflows pay off most when everyone gets the same results, not just one power user:

- **Save your best prompts.** Once a prompt works, store it as a reusable skill or custom instruction in your AI client so the whole team runs the same play.
- **Make it a habit in the repo.** Add a project rule to your agent's instructions file, such as "before adding tests, check UI Coverage for the affected view and cover the highest-risk gaps first."

> - **Standardize the report.** Share one [App Quality configuration](/ui-coverage/configuration/overview) so every teammate's agent reads the same high-signal report, and use [`profiles`](/ui-coverage/configuration/profiles) where different run types need different rules.

Reviewer: ```suggestion
- **Focus the report.** Cypress Cloud projects share one [App Quality configuration](/ui-coverage/configuration/overview) so every teammate's agent reads the same high-signal report. Use [`profiles`](/ui-coverage/configuration/profiles) to customize the config for different kinds of runs of different team ownership within a large project.
```

Updating from "share one config" since that's not something that's avoidable by default, using profiles to adjust the report in different contexts is a better recommendation.
